extensions/ql-vscode/src/databases-ui.ts

```diff
@@ -360,6 +369,45 @@ export class DatabaseUI extends DisposableObject {
     }
   };
 
+  handleRemoveOrphanedDatabases = async (): Promise<void> => {
+    logger.log('Removing orphaned databases.');
+    let dbDirs =
+      // read directory
+      (await fs.readdir(this.storagePath, { withFileTypes: true }))
+        // remove non-directories
+        .filter(dirent => dirent.isDirectory())
+        // get the full path
+        .map(dirent => path.join(this.storagePath, dirent.name))
+        // filter databases still in workspace
+        .filter(dbDir => {
+          const dbUri = Uri.file(dbDir);
+          return this.databaseManager.databaseItems.every(item => item.databaseUri.fsPath !== dbUri.fsPath);
+        });
+
+    // filter non-databases
+    dbDirs = await asyncFilter(dbDirs, isDatabaseDirectory);
+
+    if (!dbDirs.length) {
+      logger.log('No orphaned databases found.');
+      return;
+    }
+
+    // delete
+    const failures = [] as string[];
+    await Promise.all(
+      dbDirs.map(async dbDir => {
+        try {
+          logger.log(`Deleting orphaned database '${dbDir}'.`);
+          await fs.rmdir(dbDir, { recursive: true } as any);  // typings doesn't recognize the options argument
+        } catch (e) {
+          failures.push(`${e.message}: ${dbDir}`);
+        }
+      })
+    );
+    showAndLogErrorMessage(`Failed to delete orphaned databases:\n  ${failures.join('  \n')}'. Must delete manually.`);
```

Comment:
Nit: why leading space before the newline? And since this is user-facing, perhaps give them an action, e.g. 'To delete unused databases, please remove them manually from the workspace storage folder.'

Reply:
Leading space is for indentation, to indent paths we could not remove. Though, the space should come _after_ the newline.

Also, the failures should contain the full path to the database. Maybe I will use only the basename and include the storage folder elsewhere.